File: ProblemSolve/897.cpp
/********************************
 * Increasing Order Search Tree *
 ********************************/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<int> sortBST(TreeNode* root, vector<int>& sortedBST){
        if(root == NULL){
            return {};
        }
        
        sortBST(root -> left, sortedBST);
        sortedBST.push_back(root -> val);
        sortBST(root -> right, sortedBST);
        
        return sortedBST;
    }
    
    TreeNode* increasingBST(TreeNode* root) {
        vector<int> sortedBST;
        sortBST(root, sortedBST);
        
        TreeNode* ans = new TreeNode(0);
        TreeNode* curNode = ans;
        for(int node : sortedBST){
            curNode -> right = new TreeNode(node);
            curNode = curNode -> right;
        }
        
        return ans -> right;
    }
};

File: ProblemSolve/897.cs
/********************************
 * Increasing Order Search Tree *
 ********************************/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public List<int> SortBSTIncreasingOrder(TreeNode root, List<int> bstNodes){
        if(root == null){
            return null;
        }
        
        SortBSTIncreasingOrder(root.left, bstNodes);
        bstNodes.Add(root.val);
        SortBSTIncreasingOrder(root.right, bstNodes);
        
        return bstNodes;
    }
    
    public TreeNode IncreasingBST(TreeNode root) {
        List<int> nodes = SortBSTIncreasingOrder(root, new List<int>());
        TreeNode ans = new TreeNode(0);
        TreeNode ansTraverse = ans;
        foreach(int node in nodes){
            ansTraverse.right = new TreeNode(node);
            ansTraverse = ansTraverse.right;
        }
        
        return ans.right;
    }
}